File: server/src/com/inkostilation/pong/server/network/ICommandRouter.java
package com.inkostilation.pong.server.network;

import com.inkostilation.pong.commands.AbstractRequestCommand;
import com.inkostilation.pong.engine.IEngine;
import com.inkostilation.pong.exceptions.NoEngineException;

import java.io.IOException;

public interface ICommandRouter<M> {

    void route(AbstractRequestCommand<IEngine<M>, M> command) throws IOException, NoEngineException;
}

File: server/src/com/inkostilation/pong/server/network/ICommandRouter.kt
package com.inkostilation.pong.server.network

import com.inkostilation.pong.commands.AbstractRequestCommand
import com.inkostilation.pong.commands.AbstractResponseCommand
import com.inkostilation.pong.engine.IEngine

interface ICommandRouter<M> {
    fun processCommand(command: AbstractRequestCommand<IEngine<M>, M>, marker: M): List<AbstractResponseCommand<*>>
    fun start(engines: List<IEngine<M>>)
    fun reroute(marker: M, engine: Class<out IEngine<M>>): Boolean
}